File: apps/api/src/games/games.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { eq, inArray, sql } from 'drizzle-orm';
import { ConfigService } from '@nestjs/config';
import sharp from 'sharp';
import { DatabaseService } from '@/db/database.service';
import {
  games,
  type Game,
  gameObject,
  type SyncOperation,
  artStyles as artStylesView,
  type ArtStyleValue,
  GameInsert,
} from '@workspace/api-contract';
import { IgdbService, type IgdbGame } from '@/lib/igdb.service';
import { AiService } from '@/lib/ai.service';
import { S3Service } from '@/lib/s3.service';
import { R2Service } from '@/lib/r2.service';
import type { AppConfiguration } from '@/config/configuration';
import { IMAGE_GEN_DIR, IMAGE_PROMPT_SUFFIX } from '@workspace/shared';

interface GenerateImageInput {
  igdbId: number;
  includeStoryline?: boolean;
  includeGenres?: boolean;
  includeThemes?: boolean;
  artStyle?: ArtStyleValue; // effectively artStyleValue, not renaming this to be consistent with imageGen
}

@Injectable()
export class GamesService {
  constructor(
    private readonly databaseService: DatabaseService,
    private readonly igdbService: IgdbService,
    private readonly aiService: AiService,
    private readonly s3Service: S3Service,
    private readonly configService: ConfigService<AppConfiguration>,
    private readonly r2Service: R2Service,
  ) {}

  async getGameByIgdbId(igdbId: number): Promise<Game | null> {
    const [game] = await this.databaseService.db
      .select(gameObject)
      .from(games)
      .where(eq(games.igdbId, igdbId))
      .limit(1);

    return game || null;
  }

  private refreshTimeout: NodeJS.Timeout | null = null;
  private pendingRefresh: Promise<void> | null = null;

  async refreshAllGamesView(immediate = false) {
    if (this.refreshTimeout) {
      clearTimeout(this.refreshTimeout);
      this.refreshTimeout = null;
    }

    if (immediate) {
      await this.performRefresh();

      return;
    }

    return new Promise<void>((resolve) => {
      this.refreshTimeout = setTimeout(() => {
        void this.performRefresh().then(() => resolve());
      }, 1000);
    });
  }

  private async performRefresh() {
    if (this.pendingRefresh) {
      return this.pendingRefresh;
    }

    this.pendingRefresh = (async () => {
      try {
        await this.databaseService.db.execute(
          sql`REFRESH MATERIALIZED VIEW all_games`,
        );
      } catch (e) {
        console.error('Failed to refresh materialized view', e);
      } finally {
        this.pendingRefresh = null;
      }
    })();

    return this.pendingRefresh;
  }

  async syncGameByIgdbId(
    igdbId: number,
    shouldRefresh = true,
  ): Promise<{
    game: Game;
    operation: SyncOperation;
  } | null> {
    const igdbGame = await this.igdbService.getGameById(igdbId);

    if (!igdbGame) {
      return null;
    }

    const gameData = this.mapIgdbToGame(igdbGame);

    const [existingGame] = await this.databaseService.db
      .select()
      .from(games)
      .where(eq(games.igdbId, igdbId))
      .limit(1);

    if (existingGame) {
      const [updatedGame] = await this.databaseService.db
        .update(games)
        .set({
          ...gameData,
        })
        .where(eq(games.igdbId, igdbId))
        .returning();

      if (shouldRefresh) {
        void this.refreshAllGamesView();
      }

      return {
        game: updatedGame as unknown as Game,
        operation: 'updated',
      };
    }

    const [newGame] = await this.databaseService.db
      .insert(games)
      .values(gameData)
      .returning();

    if (shouldRefresh) {
      void this.refreshAllGamesView();
    }

    return {
      game: newGame as unknown as Game,
      operation: 'created',
    };
  }

  async updateGame(
    id: number,
    updates: Partial<GameInsert>,
    shouldRefresh = true,
  ): Promise<Game | null> {
    const [updatedGame] = await this.databaseService.db
      .update(games)
      .set({
        ...updates,
        updatedAt: new Date(),
      })
      .where(eq(games.id, id))
      .returning();

    if (updatedGame && shouldRefresh) {
      void this.refreshAllGamesView();
    }

    return updatedGame || null;
  }

  async deleteGame(id: number, shouldRefresh = true): Promise<number | null> {
    const [deletedGame] = await this.databaseService.db
      .delete(games)
      .where(eq(games.id, id))
      .returning({ id: games.id });

    if (deletedGame && shouldRefresh) {
      void this.refreshAllGamesView();
    }

    return deletedGame?.id ?? null;
  }

  async deleteGames(ids: number[], shouldRefresh = true): Promise<number[]> {
    const deletedRows = await this.databaseService.db
      .delete(games)
      .where(inArray(games.id, ids))
      .returning({ id: games.id });

    if (deletedRows.length > 0 && shouldRefresh) {
      void this.refreshAllGamesView();
    }

    return deletedRows.map((row) => row.id);
  }

  async validateGameForAdd(igdbId: number): Promise<{
    igdbId: number;
    existsOnIgdb: boolean;
    alreadyInDb: boolean;
    gameName: string | null;
    canAdd: boolean;
  }> {
    const [existingRow] = await this.databaseService.db
      .select({ id: games.id, name: games.name })
      .from(games)
      .where(eq(games.igdbId, igdbId))
      .limit(1);

    if (existingRow) {
      return {
        igdbId,
        existsOnIgdb: true,
        alreadyInDb: true,
        gameName: existingRow.name ?? null,
        canAdd: false,
      };
    }

    try {
      const igdbGame = await this.igdbService.getGameById(igdbId);

      if (!igdbGame) {
        return {
          igdbId,
          existsOnIgdb: false,
          alreadyInDb: false,
          gameName: null,
          canAdd: false,
        };
      }

      return {
        igdbId,
        existsOnIgdb: true,
        alreadyInDb: false,
        gameName: igdbGame.name ?? null,
        canAdd: true,
      };
    } catch {
      return {
        igdbId,
        existsOnIgdb: false,
        alreadyInDb: false,
        gameName: null,
        canAdd: false,
      };
    }
  }

  private mapIgdbToGame(igdbGame: IgdbGame): GameInsert {
    const formatUrl = (url?: string) => {
      if (!url) {
        return undefined;
      }

      const fullUrl = url.startsWith('//') ? `https:${url}` : url;

      return fullUrl.replace('t_thumb', 't_720p');
    };

    return {
      igdbId: igdbGame.id,
      name: igdbGame.name,
      summary: igdbGame.summary,
      storyline: igdbGame.storyline,
      firstReleaseDate: igdbGame.first_release_date,
      imageUrl: formatUrl(igdbGame.cover?.url),
      artworks: igdbGame.artworks?.map((art) => ({
        ...art,
        url: formatUrl(art.url),
      })),
      platforms: igdbGame.platforms?.map((p) => p.name),
      genres: igdbGame.genres?.map((g) => g.name),
      themes: igdbGame.themes?.map((t) => t.name),
      gameModes: igdbGame.game_modes?.map((m) => m.name),
      playerPerspectives: igdbGame.player_perspectives?.map((p) => p.name),
      gameEngines: igdbGame.game_engines?.map((e) => e.name),
      involvedCompanies: igdbGame.involved_companies?.map((c) => ({
        name: c.company?.name,
        developer: c.developer,
        publisher: c.publisher,
      })),
      keywords: igdbGame.keywords?.map((k) => k.name),
      franchises: igdbGame.franchises?.map((f) => f.name),
      releaseDates: igdbGame.release_dates?.map((rd) => ({
        date: rd.date,
        platform: rd.platform?.name,
      })),
    };
  }

  async generateImage(
    input: GenerateImageInput,
  ): Promise<{ success: boolean; url: string; data: Game } | null> {
    const {
      igdbId,
      includeStoryline,
      includeGenres,
      includeThemes,
      artStyle: artStyleValue, // effectively artStyleValue, not renaming this to be consistent with imageGen
    } = input;
    const game = await this.getGameByIgdbId(igdbId);
    const artStyles = await this.databaseService.db
      .select()
      .from(artStylesView);

    if (!game || !artStyleValue) {
      return null;
    }

    const artStyleDescription = artStyles.find(
      (artStyle) =>
        artStyle.value.toLowerCase() === artStyleValue.toLowerCase(),
    )?.description;
    const prompt = this.buildImagePrompt(
      game,
      {
        includeStoryline: includeStoryline ?? false,
        includeGenres: includeGenres ?? false,
        includeThemes: includeThemes ?? false,
      },
      artStyleDescription!,
    );
    const rawBuffer = await this.aiService.generateImage(prompt);
    const imageBuffer = await sharp(rawBuffer).jpeg({ quality: 85 }).toBuffer();
    const timestamp = Date.now();
    const key = `${IMAGE_GEN_DIR}/${igdbId}_${timestamp}.jpg`;

    await this.s3Service.uploadImage(key, imageBuffer, 'image/jpeg');

    const publicUrl = `${this.r2Service.r2PublicUrl}/${key}`;
    const list = Array.isArray(game.imageGen)
      ? JSON.parse(JSON.stringify(game.imageGen))
      : [];
    const newItem = {
      [artStyleValue]: {
        url: publicUrl,
        prompt: prompt,
        provider: 'cloudflare',
      },
    };
    const existingIndex = list.findIndex(
      (item: any) => item && typeof item === 'object' && artStyleValue in item,
    );

    if (existingIndex >= 0) {
      list[existingIndex] = newItem;
    } else {
      list.push(newItem);
    }

    const updatedGame = await this.updateGame(game.id, {
      aiImageUrl: publicUrl,
      aiPrompt: prompt,
      imageGen: list,
    });

    if (!updatedGame) {
      throw new NotFoundException('Failed to update game record');
    }

    return { success: true, url: publicUrl, data: updatedGame };
  }

  private buildImagePrompt(
    game: Pick<
      Game,
      'name' | 'summary' | 'storyline' | 'keywords' | 'genres' | 'themes'
    >,
    options: {
      includeStoryline: boolean;
      includeGenres: boolean;
      includeThemes: boolean;
    },
    artStyleDescription: string,
  ): string {
    const parts: string[] = [];

    parts.push(
      `${artStyleDescription} of iconic characters from "${game.name}" set within the game's distinct world`,
    );

    if (game.summary) {
      parts.push(game.summary);
    }

    if (options.includeStoryline && game.storyline) {
      parts.push(game.storyline);
    }

    if (
      options.includeGenres &&
      Array.isArray(game.genres) &&
      game.genres.length > 0
    ) {
      parts.push(`Genre: ${(game.genres as string[]).join(', ')}`);
    }

    if (
      options.includeThemes &&
      Array.isArray(game.themes) &&
      game.themes.length > 0
    ) {
      parts.push(`Themes: ${(game.themes as string[]).join(', ')}`);
    }

    if (Array.isArray(game.keywords) && game.keywords.length > 0) {
      parts.push(`Keywords: ${(game.keywords as string[]).join(', ')}`);
    }

    parts.push(IMAGE_PROMPT_SUFFIX);

    return parts.join('. ');
  }
}

File: apps/newapi/services/games_service.go
package services

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"strings"

	"gaeldle/newapi/models"
)

type GamesService struct {
	db *sql.DB
}

func NewGamesService(db *sql.DB) *GamesService {
	return &GamesService{db: db}
}

// GetGameByIgdbId retrieves a single game by its IGDB ID
func (s *GamesService) GetGameByIgdbId(igdbID int) (*models.Game, error) {
	query := `
		SELECT id, igdb_id, name, image_url, ai_image_url, ai_prompt, 
		       image_gen, info_gen, artworks, keywords, franchises, 
		       game_engines, game_modes, genres, involved_companies, 
		       platforms, player_perspectives, release_dates, themes, 
		       first_release_date, summary, storyline, created_at, updated_at
		FROM game 
		WHERE igdb_id = $1 
		LIMIT 1`

	var game models.Game
	var imageGen, infoGen, artworks, keywords, franchises, gameEngines, gameModes, genres, involvedCompanies, platforms, playerPerspectives, releaseDates, themes []byte

	err := s.db.QueryRow(query, igdbID).Scan(
		&game.ID, &game.IgdbID, &game.Name, &game.ImageURL, &game.AiImageURL, &game.AiPrompt,
		&imageGen, &infoGen, &artworks, &keywords, &franchises,
		&gameEngines, &gameModes, &genres, &involvedCompanies,
		&platforms, &playerPerspectives, &releaseDates, &themes,
		&game.FirstReleaseDate, &game.Summary, &game.Storyline, &game.CreatedAt, &game.UpdatedAt,
	)

	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	game.ImageGen = json.RawMessage(imageGen)
	game.InfoGen = json.RawMessage(infoGen)
	game.Artworks = json.RawMessage(artworks)
	game.Keywords = json.RawMessage(keywords)
	game.Franchises = json.RawMessage(franchises)
	game.GameEngines = json.RawMessage(gameEngines)
	game.GameModes = json.RawMessage(gameModes)
	game.Genres = json.RawMessage(genres)
	game.InvolvedCompanies = json.RawMessage(involvedCompanies)
	game.Platforms = json.RawMessage(platforms)
	game.PlayerPerspectives = json.RawMessage(playerPerspectives)
	game.ReleaseDates = json.RawMessage(releaseDates)
	game.Themes = json.RawMessage(themes)

	return &game, nil
}

// GetPaginatedGames retrieves games with pagination, search, and sorting
func (s *GamesService) GetPaginatedGames(page, pageSize int, q, igdbIdFilter, sortBy, sortDir string) ([]*models.Game, int, error) {
	offset := (page - 1) * pageSize

	baseQuery := `
		SELECT id, igdb_id, name, image_url, ai_image_url, ai_prompt, 
		       image_gen, info_gen, artworks, keywords, franchises, 
		       game_engines, game_modes, genres, involved_companies, 
		       platforms, player_perspectives, release_dates, themes, 
		       first_release_date, summary, storyline, created_at, updated_at
		FROM game`

	countQuery := `SELECT count(*) FROM game`

	var conditions []string
	var args []interface{}
	argIndex := 1

	if q != "" {
		conditions = append(conditions, fmt.Sprintf("name ILIKE $%d", argIndex))
		args = append(args, "%"+q+"%")
		argIndex++
	}

	if igdbIdFilter != "" {
		conditions = append(conditions, fmt.Sprintf("igdb_id::text ILIKE $%d", argIndex))
		args = append(args, "%"+igdbIdFilter+"%")
		argIndex++
	}

	if len(conditions) > 0 {
		whereClause := " WHERE " + strings.Join(conditions, " AND ")
		baseQuery += whereClause
		countQuery += whereClause
	}

	// First query the total count
	var total int
	err := s.db.QueryRow(countQuery, args...).Scan(&total)
	if err != nil {
		return nil, 0, err
	}

	// Now build sorting
	var orderBy string
	if q != "" {
		orderBy = fmt.Sprintf("ORDER BY similarity(name, $%d) DESC", argIndex)
		args = append(args, q)
		argIndex++
	} else {
		sortCol := "name"
		switch sortBy {
		case "firstReleaseDate":
			sortCol = "first_release_date"
		case "createdAt":
			sortCol = "created_at"
		case "igdbId":
			sortCol = "igdb_id"
		}

		dir := "ASC"
		if strings.ToLower(sortDir) == "desc" {
			dir = "DESC"
		}

		if sortBy == "firstReleaseDate" {
			orderBy = fmt.Sprintf("ORDER BY %s %s NULLS LAST", sortCol, dir)
		} else {
			orderBy = fmt.Sprintf("ORDER BY %s %s", sortCol, dir)
		}
	}

	baseQuery += " " + orderBy
	baseQuery += fmt.Sprintf(" LIMIT $%d OFFSET $%d", argIndex, argIndex+1)
	args = append(args, pageSize, offset)

	rows, err := s.db.Query(baseQuery, args...)
	if err != nil {
		return nil, 0, err
	}
	defer rows.Close()

	var games []*models.Game
	for rows.Next() {
		var game models.Game
		var imageGen, infoGen, artworks, keywords, franchises, gameEngines, gameModes, genres, involvedCompanies, platforms, playerPerspectives, releaseDates, themes []byte

		err := rows.Scan(
			&game.ID, &game.IgdbID, &game.Name, &game.ImageURL, &game.AiImageURL, &game.AiPrompt,
			&imageGen, &infoGen, &artworks, &keywords, &franchises,
			&gameEngines, &gameModes, &genres, &involvedCompanies,
			&platforms, &playerPerspectives, &releaseDates, &themes,
			&game.FirstReleaseDate, &game.Summary, &game.Storyline, &game.CreatedAt, &game.UpdatedAt,
		)
		if err != nil {
			return nil, 0, err
		}

		game.ImageGen = json.RawMessage(imageGen)
		game.InfoGen = json.RawMessage(infoGen)
		game.Artworks = json.RawMessage(artworks)
		game.Keywords = json.RawMessage(keywords)
		game.Franchises = json.RawMessage(franchises)
		game.GameEngines = json.RawMessage(gameEngines)
		game.GameModes = json.RawMessage(gameModes)
		game.Genres = json.RawMessage(genres)
		game.InvolvedCompanies = json.RawMessage(involvedCompanies)
		game.Platforms = json.RawMessage(platforms)
		game.PlayerPerspectives = json.RawMessage(playerPerspectives)
		game.ReleaseDates = json.RawMessage(releaseDates)
		game.Themes = json.RawMessage(themes)

		games = append(games, &game)
	}

	return games, total, nil
}

// GetRandomGames retrieves random games meeting specific constraints
func (s *GamesService) GetRandomGames(excludeIds []int, mode string, count int) ([]*models.Game, error) {
	baseQuery := `
		SELECT id, igdb_id, name, image_url, ai_image_url, ai_prompt, 
		       image_gen, info_gen, artworks, keywords, franchises, 
		       game_engines, game_modes, genres, involved_companies, 
		       platforms, player_perspectives, release_dates, themes, 
		       first_release_date, summary, storyline, created_at, updated_at
		FROM game`

	var conditions []string
	var args []interface{}
	argIndex := 1

	if len(excludeIds) > 0 {
		var placeholders []string
		for _, id := range excludeIds {
			placeholders = append(placeholders, fmt.Sprintf("$%d", argIndex))
			args = append(args, id)
			argIndex++
		}
		conditions = append(conditions, fmt.Sprintf("id NOT IN (%s)", strings.Join(placeholders, ", ")))
	}

	switch mode {
	case "artwork":
		conditions = append(conditions, "artworks IS NOT NULL AND json_array_length(artworks) > 0")
	case "cover-art":
		conditions = append(conditions, "image_url IS NOT NULL")
	case "image-gen":
		conditions = append(conditions, "image_gen IS NOT NULL AND json_array_length(image_gen) > 0")
	case "timeline", "timeline-2":
		conditions = append(conditions, "first_release_date IS NOT NULL")
	}

	if len(conditions) > 0 {
		baseQuery += " WHERE " + strings.Join(conditions, " AND ")
	}

	baseQuery += fmt.Sprintf(" ORDER BY RANDOM() LIMIT $%d", argIndex)
	args = append(args, count)

	rows, err := s.db.Query(baseQuery, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var games []*models.Game
	for rows.Next() {
		var game models.Game
		var imageGen, infoGen, artworks, keywords, franchises, gameEngines, gameModes, genres, involvedCompanies, platforms, playerPerspectives, releaseDates, themes []byte

		err := rows.Scan(
			&game.ID, &game.IgdbID, &game.Name, &game.ImageURL, &game.AiImageURL, &game.AiPrompt,
			&imageGen, &infoGen, &artworks, &keywords, &franchises,
			&gameEngines, &gameModes, &genres, &involvedCompanies,
			&platforms, &playerPerspectives, &releaseDates, &themes,
			&game.FirstReleaseDate, &game.Summary, &game.Storyline, &game.CreatedAt, &game.UpdatedAt,
		)
		if err != nil {
			return nil, err
		}

		game.ImageGen = json.RawMessage(imageGen)
		game.InfoGen = json.RawMessage(infoGen)
		game.Artworks = json.RawMessage(artworks)
		game.Keywords = json.RawMessage(keywords)
		game.Franchises = json.RawMessage(franchises)
		game.GameEngines = json.RawMessage(gameEngines)
		game.GameModes = json.RawMessage(gameModes)
		game.Genres = json.RawMessage(genres)
		game.InvolvedCompanies = json.RawMessage(involvedCompanies)
		game.Platforms = json.RawMessage(platforms)
		game.PlayerPerspectives = json.RawMessage(playerPerspectives)
		game.ReleaseDates = json.RawMessage(releaseDates)
		game.Themes = json.RawMessage(themes)

		games = append(games, &game)
	}

	return games, nil
}

// SearchGames performs a similarity search on games
func (s *GamesService) SearchGames(q string, limit int, mode string) ([]*models.Game, error) {
	baseQuery := `
		SELECT id, igdb_id, name, image_url, ai_image_url, ai_prompt, 
		       image_gen, info_gen, artworks, keywords, franchises, 
		       game_engines, game_modes, genres, involved_companies, 
		       platforms, player_perspectives, release_dates, themes, 
		       first_release_date, summary, storyline, created_at, updated_at
		FROM game`

	var conditions []string
	var args []interface{}
	argIndex := 1

	conditions = append(conditions, fmt.Sprintf("name ILIKE $%d", argIndex))
	args = append(args, "%"+q+"%")
	argIndex++

	switch mode {
	case "artwork":
		conditions = append(conditions, "artworks IS NOT NULL AND json_array_length(artworks) > 0")
	case "cover-art":
		conditions = append(conditions, "image_url IS NOT NULL")
	case "image-gen":
		conditions = append(conditions, "image_gen IS NOT NULL AND json_array_length(image_gen) > 0")
	case "timeline", "timeline-2":
		conditions = append(conditions, "first_release_date IS NOT NULL")
	}

	baseQuery += " WHERE " + strings.Join(conditions, " AND ")
	baseQuery += fmt.Sprintf(" ORDER BY similarity(name, $%d) DESC LIMIT $%d", argIndex, argIndex+1)
	args = append(args, q, limit)

	rows, err := s.db.Query(baseQuery, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var games []*models.Game
	for rows.Next() {
		var game models.Game
		var imageGen, infoGen, artworks, keywords, franchises, gameEngines, gameModes, genres, involvedCompanies, platforms, playerPerspectives, releaseDates, themes []byte

		err := rows.Scan(
			&game.ID, &game.IgdbID, &game.Name, &game.ImageURL, &game.AiImageURL, &game.AiPrompt,
			&imageGen, &infoGen, &artworks, &keywords, &franchises,
			&gameEngines, &gameModes, &genres, &involvedCompanies,
			&platforms, &playerPerspectives, &releaseDates, &themes,
			&game.FirstReleaseDate, &game.Summary, &game.Storyline, &game.CreatedAt, &game.UpdatedAt,
		)
		if err != nil {
			return nil, err
		}

		game.ImageGen = json.RawMessage(imageGen)
		game.InfoGen = json.RawMessage(infoGen)
		game.Artworks = json.RawMessage(artworks)
		game.Keywords = json.RawMessage(keywords)
		game.Franchises = json.RawMessage(franchises)
		game.GameEngines = json.RawMessage(gameEngines)
		game.GameModes = json.RawMessage(gameModes)
		game.Genres = json.RawMessage(genres)
		game.InvolvedCompanies = json.RawMessage(involvedCompanies)
		game.Platforms = json.RawMessage(platforms)
		game.PlayerPerspectives = json.RawMessage(playerPerspectives)
		game.ReleaseDates = json.RawMessage(releaseDates)
		game.Themes = json.RawMessage(themes)

		games = append(games, &game)
	}

	return games, nil
}

// SyncGameByIgdbId dummy implementation
func (s *GamesService) SyncGameByIgdbId(igdbID int, shouldRefresh bool, actorID string) (map[string]interface{}, error) {
	return map[string]interface{}{
		"success":   true,
		"message":   "Game created",
		"operation": "created",
		"data": map[string]interface{}{
			"id":     9999,
			"igdbId": igdbID,
			"name":   "Dummy Game Name",
		},
	}, nil
}

// UpdateGame dummy implementation
func (s *GamesService) UpdateGame(id int, updates map[string]interface{}) (map[string]interface{}, error) {
	return map[string]interface{}{
		"success": true,
		"data": map[string]interface{}{
			"id":     id,
			"name":   "Dummy Updated Game Name",
			"igdbId": 1234,
		},
	}, nil
}

// DeleteGame dummy implementation
func (s *GamesService) DeleteGame(id int) (int, error) {
	return id, nil
}

// DeleteGames dummy implementation
func (s *GamesService) DeleteGames(ids []int) ([]int, error) {
	return ids, nil
}

// ValidateGameForAdd dummy implementation
func (s *GamesService) ValidateGameForAdd(igdbID int, actorID string) (map[string]interface{}, error) {
	return map[string]interface{}{
		"igdbId":       igdbID,
		"existsOnIgdb": true,
		"alreadyInDb":  false,
		"gameName":     "Dummy Candidate",
		"canAdd":       true,
	}, nil
}

// GenerateImage dummy implementation
func (s *GamesService) GenerateImage(input interface{}, actorID string) (map[string]interface{}, error) {
	return map[string]interface{}{
		"success":   true,
		"messageId": "dummy-msg-id-123",
	}, nil
}

// BuildImagePrompt dummy implementation
func (s *GamesService) BuildImagePrompt(game interface{}, options interface{}, artStyleDescription string) string {
	return "dummy prompt"
}

// RefreshAllGamesView dummy implementation
func (s *GamesService) RefreshAllGamesView(immediate bool) error {
	return nil
}

// RefreshQueriedGamesView dummy implementation
func (s *GamesService) RefreshQueriedGamesView() error {
	return nil
}
